### ISSUE TITLE
`pin_subpackage` fails when package declared further down is referenced

### ISSUE THREAD

**Issue by borchero** [prefix-dev/rattler-build#1245]
Using rattler-build from `main` on `darwin-aarch64`, the following recipe where `test1` is referenced from `test2` but declared after

```yaml
recipe:
  name: test-split
  version: v1

build:
  number: 0
  noarch: python

source:
  path: ./

outputs:
  - package:
      name: test2
    requirements:
      run:
        - ${{ pin_subpackage('test1', exact=True) }}
  - package:
      name: test1
```

results in a stackoverflow:

```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
[1]    82318 abort      '/Users/borchero/Downloads/rattler-build-aarch64-apple-darwin 2/rattler-build
```

---

When declaring `test1` before `test2`, the build works.

**Comment by wolfv** [comment]
Thanks! I'll take a look right now.

**Comment by borchero** [comment]
Let me know if I should test some build on the recipe that prompted my last few issues :smile:

**Comment by wolfv** [comment]
Fixed in #1250
